include/mbgl/style/function/type.hpp

```diff
@@ -0,0 +1,117 @@
+#pragma once
+
+#include <vector>
+#include <mbgl/util/optional.hpp>
+#include <mbgl/util/variant.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+namespace type {
+
+class Primitive;
+class Variant;
+class Array;
+class NArgs;
+class Typename;
+class Lambda;
+
+using ValueType = variant<
```

Comment:
Ah, another thought: i think we should allow `Typename`s to be considered a valid expression type, so that an `Expression` can represent an un-typechecked AST. (E.g. before we do type inference, the type of `[==, 1, 2]` would be `T`.)